akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/AroundRequestDirectivesSpec.scala

```diff
@@ -0,0 +1,220 @@
+  }
+
+  var results = Map.empty[String, TestResult]
+
+  // Run route method for testing Request-Timeout scenarios
+  def runRoute(route: Route, routePath: String): HttpResponse = {
+    val (hostname, port) = SocketUtil.temporaryServerHostnameAndPort()
+    val binding = Http().bindAndHandle(route, hostname, port)
+    val response = Http().singleRequest(HttpRequest(uri = s"http://$hostname:$port/$routePath")).futureValue
+    binding.flatMap(_.unbind()).futureValue
+    response.discardEntityBytes()
+    response
+  }
+
+  "the aroundRequest directive wrapping an unsealed route" should {
+    "invoke test function on success" in {
+      val route: Route =
+        aroundRequest(behaviour) {
+          entity(as[String]) { _ =>
+            completeOk
+          }
+        }
+      runRoute(route, "test").status should ===(StatusCodes.OK)
+      eventually {
+        results("/test") shouldBe TestComplete(StatusCodes.OK)
+      }
+    }
+    "invoke test function on reject" in {
+      val route: Route =
+        aroundRequest(behaviour) {
+          entity(as[String]) { _ =>
+            reject(MethodRejection(HttpMethods.GET))
+          }
+        }
+      runRoute(route, "test").status should ===(StatusCodes.MethodNotAllowed)
+      eventually {
+        results("/test") shouldBe TestRejection
+      }
+    }
+    "invoke test function on failure" in {
+      val route: Route =
+        aroundRequest(behaviour) {
+          entity(as[String]) { _ =>
+            fail(new Exception())
+          }
+        }
+      runRoute(route, "test").status should ===(StatusCodes.InternalServerError)
+      eventually {
+        results("/test") shouldBe TestUnknown
+      }
+    }
+    "invoke test function on an exception" in {
+      val route: Route =
+        aroundRequest(behaviour) {
+          entity(as[String]) { _ =>
+            throw new Exception()
+          }
+        }
+      runRoute(route, "test").status should ===(StatusCodes.InternalServerError)
+      eventually {
+        results("/test") shouldBe TestUnknown
+      }
+    }
+    "handle timeouts" in {
+      def slowFuture(): Future[String] = Promise[String].future
+      val route: Route =
+        aroundRequest(behaviour) {
+          withRequestTimeout(1.seconds) { // modifies the global akka.http.server.request-timeout for this request
+            val response: Future[String] = slowFuture() // very slow
+            complete(response)
+          }
+        }
+      runRoute(route, "test").status should ===(StatusCodes.ServiceUnavailable)
+      eventually {
+        results("/test") shouldBe TestComplete(StatusCodes.ServiceUnavailable)
+      }
+    }
+  }
+
+  "the aroundRequest directive when used on a chunked response" should {
+    "call the onDone function when last bytes are sent" in {
+
+      val timed: RequestContext => Try[RouteResult] => Unit = ctx => {
+        val testKey = ctx.request.uri.toRelative.toString
+        val start = System.currentTimeMillis()
```

Comment:
Shouldn't this use `System.nanoTime()`? Unlike `currentTimeMillis` it's guaranteed to be monotonic.